packages/dev/core/src/Physics/v2/ragdoll.ts

```diff
@@ -0,0 +1,331 @@
+                distanceFromParentBoxToBone,
+                myConnectedPivot,
+                this._boxConfigs[i].rotationAxis!,
+                this._boxConfigs[i].rotationAxis!,
+                this._scene
+            );
+            this._aggregates[boneParentIndex].body.addConstraint(this._aggregates[i].body, joint);
+        }
+    }
+
+    private _syncBonesAndBoxes(): void {
+        if (this._pauseSync) {
+            return;
+        }
+
+        if (this._ragdollMode) {
+            const rootBoneDirection = this._bones[this._rootBoneIndex].getDirection(this._boxConfigs[this._rootBoneIndex].boneOffsetAxis as any, this._mesh);
+            const rootBoneOffsetPosition = this._bones[this._rootBoneIndex]
+                .getAbsolutePosition(this._mesh)
+                .add(rootBoneDirection.scale(this._boxConfigs[this._rootBoneIndex].boxOffset as any));
+
+            this._bones[this._rootBoneIndex].setAbsolutePosition(this._boxes[this._rootBoneIndex].position, this._mesh);
+            this._addImpostorRotationToBone(this._rootBoneIndex);
+
+            const rootPos = this._aggregates[this._rootBoneIndex].body.transformNode.position;
+
+            // Move the mesh, to guarantee alignment between root bone and impostor box position
+            const dist = rootBoneOffsetPosition.subtract(rootPos);
+            this._mesh.position = this._mesh.position.subtract(dist);
+
+            for (let i = 0; i < this._bones.length; i++) {
+                if (i == this._rootBoneIndex) continue;
+                this._addImpostorRotationToBone(i);
+            }
+        }
+    }
+
+    private _addImpostorRotationToBone(boneIndex: number): void {
+        const newRotQuat = this._aggregates[boneIndex].body?.transformNode?.rotationQuaternion?.multiply(this._initialRotation[boneIndex]);
+        if (newRotQuat) {
+            this._bones[boneIndex].setRotationQuaternion(newRotQuat, Space.WORLD, this._mesh);
+        }
+    }
+
+    // Return true if root bone is valid/exists in this.bonesNames. false otherwise.
+    private _defineRootBone(): boolean {
+        const skeletonRoots = this._skeleton.getChildren();
+        if (skeletonRoots.length != 1) {
+            console.log("Ragdoll creation failed: there can only be one root in the skeleton.");
+            return false;
+        }
+
+        this._rootBoneName = skeletonRoots[0].name;
+        this._rootBoneIndex = this._boneNames.indexOf(this._rootBoneName);
+        if (this._rootBoneIndex == -1) {
+            console.log("Ragdoll creation failed: the array boneNames doesn't have the root bone. The root bone is", this._skeleton.getChildren());
+            return false;
+        }
+
+        return true;
+    }
+
+    private _findNearestParent(boneIndex: number): any {
+        let nearestParent: Nullable<Bone> | undefined = this._bones[boneIndex].getParent();
+        do {
+            if (nearestParent != null && (this._boneNames as any).includes(nearestParent.name)) {

```

Comment:
are we missing the declaration for Array.prototype.includes?

Reply:
Fixed.